clients/venice-producer/src/main/java/com/linkedin/venice/producer/AbstractVeniceProducer.java

```diff
@@ -0,0 +1,425 @@
+    } else if (object instanceof CharSequence) { // convenience option.
+      return STRING_SCHEMA;
+    } else if (object instanceof Integer) {
+      return INT_SCHEMA;
+    } else if (object instanceof Long) {
+      return LONG_SCHEMA;
+    } else if (object instanceof Double) {
+      return DOUBLE_SCHEMA;
+    } else if (object instanceof Float) {
+      return FLOAT_SCHEMA;
+    } else if (object instanceof byte[] || object instanceof ByteBuffer) {
+      return BYTES_SCHEMA;
+    } else if (object instanceof Boolean) {
+      return BOOL_SCHEMA;
+    } else {
+      throw new VeniceException(
+          "Venice Producer only supports Avro objects, and primitives, found object of class: "
+              + object.getClass().toString());
+    }
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(K key, V value) {
+    return asyncPutInternal(APP_DEFAULT_LOGICAL_TS, key, value);
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(long logicalTime, K key, V value) {
+    if (logicalTime < 0) {
+      return getFutureCompletedExceptionally("Logical time must be a non-negative value. Got: " + logicalTime);
+    }
+
+    return asyncPutInternal(logicalTime, key, value);
+  }
+
+  private CompletableFuture<DurableWrite> asyncPutInternal(long logicalTime, K key, V value) {
+    String error = validateProducer();
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
+    }
+
+    producerMetrics.recordPutRequest();
+    return CompletableFuture.supplyAsync(() -> {
+      Schema valueSchema;
+      try {
+        valueSchema = getSchemaFromObject(value);
+      } catch (Exception e) {
+        producerMetrics.recordFailedRequest();
+        throw e;
+      }
+      // Might block
+      int valueSchemaId;
+      Exception schemaReadException = null;
+      try {
+        valueSchemaId = schemaReader.getValueSchemaId(valueSchema);
+      } catch (Exception e) {
+        valueSchemaId = SchemaData.INVALID_VALUE_SCHEMA_ID;
+        schemaReadException = e;
+      }
+      if (valueSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID) {
+        producerMetrics.recordFailedRequest();
+        throw new VeniceException(
+            "Could not find a registered schema id for schema: " + valueSchema
+                + ". This might be transient if the schema has been registered recently.",
+            schemaReadException);
+      }
+      final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+      final Instant sendStartTime = Instant.now();
+      final PubSubProducerCallback callback = (PubSubProduceResult produceResult, Exception exception) -> {
+        Duration sendDuration = Duration.between(sendStartTime, Instant.now());
+        if (exception == null) {
+          producerMetrics.recordSuccessfulRequestWithLatency(sendDuration.toMillis());
+          completableFuture.complete(null);
+        } else {
+          producerMetrics.recordFailedRequest();
```

Comment:
can we add a log here?

Reply:
Yeah, I can. Do you think this will flood the logs? I guess it's okay, though